### PR TITLE
[RLlib] Replace Learner/LearnerGroup specific config classes (e.g. `LearnerHyperparameters`) with `AlgorithmConfig`.

### PR DESCRIPTION
**Pull request by sven1977**
This PR implements some larger cleanups on RLlib's new API stack:

- It addresses some circular import problems and generally tries to:
- - Always allow importing higher level classes (e.g. AlgorithmConfig) from within lower level classes (e.g. PPOTorchLearner).
- - Make sure that higher level classes (e.g. AlgorithmConfig) only import lower level ones if absolutely necessary and if so try to import these either in-line OR during TYPE_CHECKING only.

- It replaces all Learner/LearnerGroup specific config classes (e.g. `LearnerHyperparameters`) by simply using `AlgorithmConfig`. AlgorithmConfig already today contains all the information necessary to build these respective algo sub-components, such as LearnerGroup or Learner.
- All Learner/LearnerGroup APIs that used to take `LearnerHyperparameter` as arguments have been changed to accept AlgorithmConfig objects instead (e.g. `Learner.compute_loss_for_module`).
- All `LearnerHyperparameter` classes have been removed.
- `LearnerGroupConfig` has been removed.
- `ScalingConfig`, `FrameworkHyperparameters`, and `LearnerSpec` all have been removed in favor of `AlgorithmConfig` (which already today contains all the information necessary to build the respective components).
- The construction of RLModules (single- or multi-agent) has been unified, simplified, and can now be done more easily through an algo config object:
```
# 1) Create config.
config = AlgorithmConfig().rl_module(...).training(learner_class=...).experimental(_enable_new_api_stack=True)

# 1.5) To create a module, we need to have access to an env:
env = gym.make("CartPole-v1")

# 2) single-agent RLModule:
marl_module_spec = config.get_marl_module_spec(env=env)  # <- alternatively, one can provide a `spaces` dict.
sarl_module_spec = marl_module_spec.module_specs[DEFAULT_POLICY_ID]
sarl_module = sarl_module_spec.build()

# 3) multi-agent RLModule:
marl_module_spec = config.get_marl_module_spec()  # <- `env` argument not needed here iff spaces are already specified in `config.policies` dict.
marl_module = marl_module_spec.build()
```

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/algorithm_config.py [hunk 1]
```diff
@@ -1138,6 +1152,71 @@ class directly. Note that this arg can also be specified via
             logger_creator=self.logger_creator,
         )
 
+    def build_learner_group(
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:1155]: New APIs: Learner and LearnerGroup can now be directly built from within a config object.

File: rllib/algorithms/algorithm_config.py [hunk 2]
```diff
@@ -769,14 +774,6 @@ def validate(self) -> None:
                 "`config.evaluation(enable_async_evaluation=True) on your config "
                 "object to fix this problem."
             )
-        if not (
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:772]: Moved this below into `rollouts()` method to check more immediately (this is only a single arg check after all).

File: rllib/algorithms/algorithm.py [hunk 3]
```diff
@@ -763,9 +759,16 @@ def setup(self, config: AlgorithmConfig) -> None:
             if hasattr(local_worker, "marl_module_spec"):
                 module_spec = local_worker.marl_module_spec
             else:
-                module_spec = self.config.get_marl_module_spec(policy_dict=policy_dict)
-            learner_group_config = self.config.get_learner_group_config(module_spec)
-            self.learner_group = learner_group_config.build()
+                module_spec = self.config.get_marl_module_spec(
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:762]: New way of doing things: Use config's APIs to build components.

File: rllib/algorithms/algorithm.py [hunk 4]
```diff
@@ -763,9 +759,16 @@ def setup(self, config: AlgorithmConfig) -> None:
             if hasattr(local_worker, "marl_module_spec"):
                 module_spec = local_worker.marl_module_spec
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:760]: Old way: RolloutWorker has "secret" `marl_module_spec` prop.

File: rllib/algorithms/algorithm_config.py [hunk 5]
```diff
@@ -3399,85 +3488,6 @@ def get_marl_module_spec(
 
         return marl_module_spec
 
-    def get_learner_group_config(self, module_spec: ModuleSpec) -> LearnerGroupConfig:
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:3402]: No longer needed.

File: rllib/algorithms/algorithm_config.py [hunk 6]
```diff
@@ -3399,85 +3488,6 @@ def get_marl_module_spec(
...
-                torch_compile=self.torch_compile_learner,
-                torch_compile_cfg=self.get_torch_compile_learner_config(),
-                torch_compile_what_to_compile=self.torch_compile_learner_what_to_compile,  # noqa: E501
-            )
-        elif self.framework_str == "tf2":
-            config.framework(eager_tracing=self.eager_tracing)
-
-        return config
-
-    def get_learner_hyperparameters(self) -> LearnerHyperparameters:
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:3439]: No longer needed.

File: rllib/algorithms/appo/appo_learner.py [hunk 7]
```diff
@@ -1,42 +1,15 @@
...
 from ray.rllib.utils.schedules.scheduler import Scheduler
 
 
-LEARNER_RESULTS_KL_KEY = "mean_kl_loss"
-LEARNER_RESULTS_CURR_KL_COEFF_KEY = "curr_kl_coeff"
-OLD_ACTION_DIST_KEY = "old_action_dist"
-OLD_ACTION_DIST_LOGITS_KEY = "old_action_dist_logits"
-
-
-@dataclass
```
Comment (sven1977) [inline rllib/algorithms/appo/appo_learner.py:22]: No longer needed.

File: rllib/algorithms/impala/impala.py [hunk 8]
```diff
@@ -432,32 +429,8 @@ def validate(self) -> None:
                     f"than or equal to `train_batch_size` ({self.train_batch_size})!"
                 )
 
-    @override(AlgorithmConfig)
```
Comment (sven1977) [inline rllib/algorithms/impala/impala.py:435]: No longer needed.

File: rllib/algorithms/impala/impala.py [hunk 9]
```diff
@@ -1267,3 +1242,23 @@ def process_episodes(self, batch: SampleBatchType) -> SampleBatchType:
 
     def get_host(self) -> str:
         return platform.node()
+
+
+def _reduce_impala_results(results: List[ResultDict]) -> ResultDict:
```
Comment (sven1977) [inline rllib/algorithms/impala/impala.py:1247]: Moved here as it's only needed in this very file.

File: rllib/algorithms/impala/impala_learner.py [hunk 10]
```diff
@@ -1,52 +1,14 @@
...
-    NUM_ENV_STEPS_TRAINED,
-)
 from ray.rllib.utils.schedules.scheduler import Scheduler
-from ray.rllib.utils.typing import ResultDict
-
-
-LEARNER_RESULTS_CURR_ENTROPY_COEFF_KEY = "curr_entropy_coeff"
-
-
-@dataclass
```
Comment (sven1977) [inline rllib/algorithms/impala/impala_learner.py:26]: No longer needed.

File: rllib/core/learner/learner_group_config.py [hunk 11]
```diff
@@ -1,168 +0,0 @@
-from typing import Optional, Type, TYPE_CHECKING, Union
```
Comment (sven1977) [inline rllib/core/learner/learner_group_config.py:1]: No longer needed.

File: rllib/core/learner/tests/test_learner_group.py [hunk 12]
```diff
@@ -115,16 +128,42 @@ def setUp(self) -> None:
     def tearDown(self) -> None:
         ray.shutdown()
 
+    def test_learner_group_build_from_algorithm_config(self):
```
Comment (sven1977) [inline rllib/core/learner/tests/test_learner_group.py:131]: Added this test case.

File: rllib/core/testing/utils.py [hunk 13]
```diff
@@ -93,96 +32,16 @@ def get_optimizer_default_class(framework: str) -> Type[Optimizer]:
         raise ValueError(f"Unsupported framework: {framework}")
 
 
-@DeveloperAPI
```
Comment (sven1977) [inline rllib/core/testing/utils.py:96]: No longer needed.

File: rllib/core/testing/utils.py [hunk 14]
```diff
@@ -31,54 +18,6 @@
 DEFAULT_POLICY_ID = "default_policy"
 
 
-@DeveloperAPI
```
Comment (sven1977) [inline rllib/core/testing/utils.py:34]: No longer needed.

File: rllib/algorithms/appo/tests/test_appo_learner.py [hunk 15]
```diff
@@ -95,17 +93,9 @@ def test_appo_loss(self):
...
-                    observation_space=policy.observation_space,
-                    action_space=policy.action_space,
-                    model_config_dict=policy.config["model"],
-                    catalog_class=algo_config.rl_module_spec.catalog_class,
-                )
+            learner_group = algo_config.build_learner_group(
+                env=algo.workers.local_worker().env
             )
-            learner_group_config.num_learner_workers = 0
-            learner_group = learner_group_config.build()
```
Comment (kouroshHakha) [inline rllib/algorithms/appo/tests/test_appo_learner.py:108]: how are we setting the equivalent of `num_learner_workers = 0`?

**Review by kouroshHakha — COMMENTED** [review]
LGTM. one comment.

File: rllib/algorithms/appo/tests/test_appo_learner.py (hunk 15 above)
Comment (sven1977) [reply rllib/algorithms/appo/tests/test_appo_learner.py:108]: Now via `config.resources(num_learner_workers=0)` (which is the default anyways so it has no effect on this test here).

**Review by kouroshHakha — APPROVED** [review]

**Pull request merged**